### PR TITLE
[SPARK-14495][SQL][1.6] fix resolution failure of having clause with distinct aggregate function

### PR DESCRIPTION
**Pull request by xwu0226**
#### Symptom:

In the latest **branch 1.6**, when a `DISTINCT` aggregation function is used in the `HAVING` clause, Analyzer throws `AnalysisException` with a message like following:

```
resolved attribute(s) gid#558,id#559 missing from date#554,id#555 in operator !Expand [List(date#554, null, 0, if ((gid#558 = 1)) id#559 else null),List(date#554, id#555, 1, null)], [date#554,id#561,gid#560,if ((gid = 1)) id else null#562];
```
#### Root cause:

The problem is that the distinct aggregate in having condition are resolved by the rule `DistinctAggregationRewriter` twice, which messes up the resulted `EXPAND` operator. 

In a `ResolveAggregateFunctions` rule, when resolving `Filter(havingCondition, _: Aggregate)`, the `havingCondition` is resolved as an `Aggregate` in a nested loop of analyzer rule execution (by invoking `RuleExecutor.execute`). At this nested level of analysis, the rule `DistinctAggregationRewriter` rewrites this distinct aggregate clause to an expanded two-layer aggregation, where the `aggregateExpresssions` of the final `Aggregate` contains the resolved `gid` and the aggregate expression attributes (In the above case, they are  `gid#558, id#559`).  

After completion of the nested analyzer rule execution, the resulted `aggregateExpressions` in the `havingCondition` is pushed down into the underlying `Aggregate` operator. The `DistinctAggregationRewriter` rule is executed again. The `projections` field of `EXPAND` operator is populated with the `aggregateExpressions` of the `havingCondition` mentioned above. However, the attributes (In the above case, they are `gid#558, id#559`) in the projection list of `EXPAND` operator can not be found in the underlying relation.
#### Solution:

This PR retrofits part of [#11579](https://github.com/apache/spark/pull/11579) that moves the `DistinctAggregationRewriter` to the beginning of Optimizer, so that it guarantees that the rewrite only happens after all the aggregate functions are resolved first. Thus, it avoids resolution failure. 
#### How is the PR change tested

New [test cases ](https://github.com/xwu0226/spark/blob/f73428f94746d6d074baf6702589545bdbd11cad/sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala#L927-L988) are added to drive `DistinctAggregationRewriter` rewrites for multi-distinct aggregations , involving having clause. 

A following up PR will be submitted to add these test cases to master(2.0) branch.

### CODE REVIEW

**Comment by gatorsmile** [comment]
@rxin @cloud-fan Thanks!

**Comment by gatorsmile** [comment]
@xwu0226 Because this is for Branch 1.6 only, please update the PR title to 
```
[SPARK-14495][SQL] [1.6] Fix Resolution Failure of Having Clause with Distinct Aggregate Functions
```

**Comment by cloud-fan** [comment]
OK to test

**Comment by cloud-fan** [comment]
So #11579 accidentally fixed this bug in 2.0?

**Comment by cloud-fan** [comment]
ok to test

**Comment by xwu0226** [comment]
@cloud-fan Yes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DistinctAggregationRewriter.scala [hunk 1]
```diff
@@ -123,15 +119,7 @@ case class DistinctAggregationRewriter(conf: CatalystConf) extends Rule[LogicalP
       .filter(_.isDistinct)
       .groupBy(_.aggregateFunction.children.toSet)
 
-    val shouldRewrite = if (conf.specializeSingleDistinctAggPlanning) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DistinctAggregationRewriter.scala]: is this flag still useful in 1.6? cc @davies @yhuai

Comment (xwu0226) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DistinctAggregationRewriter.scala]: This flag is for the purpose of benchmarking the performance of single distinct aggregation by `DistinctAggregationRewriter`. The default value is false, which means `DistinctAggregationRewriter` will not be used for a single distinct case. I see 2.0 has removed this flag, so i guess the decision has been made. If it is still needed for 1.6, I can add it back, which will involves more change in Optimizer to take the CatalystConf.  Please let me know. Thanks!

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DistinctAggregationRewriter.scala]: Even it's not that useful, we should not remove it in minor release.

Comment (xwu0226) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DistinctAggregationRewriter.scala]: @davies Thanks for your input! Let me modify Optimizer to add the conf parameter.

**Comment by xwu0226** [comment]
@davies @cloud-fan  I modified the change to keep the SQLConf property. In order to pass `conf: CatalystConf` to Optimizer, I did similar thing as how 2.0 branch handles Optimizer. Please help take a look. Thanks!

**Comment by gatorsmile** [comment]
retest this please

**Comment by xwu0226** [comment]
retest this please

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala [hunk 2]
```diff
@@ -560,80 +560,73 @@ abstract class AggregationQuerySuite extends QueryTest with SQLTestUtils with Te
   }
 
   test("single distinct column set") {
-    Seq(true, false).foreach { specializeSingleDistinctAgg =>
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala]: This config is not removed, looks like we don't need to change this test?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala [hunk 3]
```diff
@@ -44,10 +44,12 @@ class PlannerSuite extends SharedSQLContext {
...
 
-    // For the new aggregation code path, there will be four aggregate operator for
-    // distinct aggregations.
+    // For the new aggregation code path, there will be three aggregate operator for
+    // distinct aggregations. There used to be four aggregate operators because single
+    // distinct aggregate used to trigger DistinctAggregationRewriter rewrite. Now the
+    // the rewrite only happens when there are multiple distinct aggregations.
     assert(
-      aggregations.size == 2 || aggregations.size == 4,
+      aggregations.size == 2 || aggregations.size == 3,
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: why this change? This PR just moves the rewrite rule to the beginning of optimizer rule right?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala (hunk 2 above)
Comment (xwu0226) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala]: yes. you are right. I will add it back.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala (hunk 3 above)
Comment (xwu0226) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: Before my change of moving the `DistinctAggregationRewriter` rule to Optimizer, Analyzer resolved the single distinct case into 2-layered non-distinct Aggregates and an EXPAND as following:
```
Aggregate [value#3], [value#3,(count(if ((gid#48 = 1)) key#49 else null),mode=Complete,isDistinct=false) AS count(key)#47L]
+- Aggregate [value#3,key#49,gid#48], [value#3,key#49,gid#48]
   +- Expand [ArrayBuffer(value#3, key#2, 1)], [value#3,key#49,gid#48]
      +- LogicalRDD [key#2,value#3], MapPartitionsRDD[3] at beforeAll at BeforeAndAfterAll.scala:187
```
Then, the resulted physical Plan in this testcase had 4 `TungstenAggregate` generated by [aggregate/utils.planAggregateWithoutDistinct](https://github.com/apache/spark/blob/branch-1.6/sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/utils.scala#L65-L131). So this testcase checked for 4 aggregations. 

After my change, there is no rewrite for the single distinct aggregate anymore, the analyzed plan looks like this:
```
Aggregate [value#3], [value#3,(count(key#2),mode=Complete,isDistinct=true) AS count(key)#47L]
+- LogicalRDD [key#2,value#3], MapPartitionsRDD[3] at beforeAll at BeforeAndAfterAll.scala:187
```
Then, the physical plan are resulted in 3 `TungstenAggregate` by [aggregate/utils.planAggregateWithOneDistinct](https://github.com/apache/spark/blob/branch-1.6/sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/utils.scala#L160-L288)

Therefore, I need to modify this testcase.

Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: Why is there no rewrite after your change? You just moved the rule but didn't change it.

Comment (xwu0226) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala]: The rewrite I mentioned was by Analyzer. In this testcase, the analyzed logical plan, instead of the optimized one, is passed to generate physical plan.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala [hunk 4]
```diff
@@ -930,6 +930,69 @@ abstract class AggregationQuerySuite extends QueryTest with SQLTestUtils with Te
         Row(11) :: Nil)
     }
   }
+
+  test("SPARK-14495: two distinct aggregation with having clause of one distinct aggregation") {
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala]: these 4 tests look a little verbose to me. The key to trigger this bug is to put distinct aggregate function in having, right?

Comment (xwu0226) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala]: Yes. Thanks for pointing it out! The original thinking was that because I removed the configuration property, single distinct aggregate case will not get into rewrite in Optimizer. Therefore, I wanted to cover cases for both single distinct and multi-distinct in having clause. I think I can just keep the first one. Any suggestions? Thanks!

Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala]: SGTM, please also update the test case name, thanks!

Comment (xwu0226) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala]: Ok. Thanks! I will update and push once the current test build is done.

**Comment by cloud-fan** [comment]
merging to 1.6, thanks!

**Comment by cloud-fan** [comment]
Can you close it? Looks like PR not merged to master won't be closed automatically.

**Comment by xwu0226** [comment]
@cloud-fan Thank! I will close it.

**Comment by Jiangwlee** [comment]
Thanks! I just encountered this issue today.
